_drafts/armv7_mpu.md

```diff
@@ -0,0 +1,376 @@
+
+### MPU Region Number Register
+
+`MPU_RNR` - Address 0xE000ED98 register bit assignments:
+![](img/armv7m_mpu/mpu_rnr.png)
+
+As we saw in the `MPU_TYPE` register above, the MPU is broken into a certain number of configurable-regions. The _M0+_, _M3_ & _M4_ can support up to 8 regions when the MPU is implemented and the M7 can support up to 16! The `REGION` field is used to select the active region we want to read or update. Once the region is selected, we can read or write `MPU_RBAR` & `MPU_RASR` to see the configuration and make changes.
+
+**Caution** It's worth mentioning that there is a priority scheme to these regions. If the same memory range is covered in multiple MPU regions, the configuration in the **highest** `REGION` number will be applied. So for example, if `REGION 0` turns off access to the first _32kB_ of RAM for _Unprivileged_ mode but `REGION 1` makes the same RAM region read-only, `REGION 1` settings will take precedence. This means code running in _Unprivileged_ mode **will** be able to read the memory.
+
+**Note** The default memory map enabled with the `PRIVDEFENA` mentioned above has a priority of `-1` and so it will be overriden by any configurations made in these regions.
+
+### MPU Region Base Address Register
+
+`MPU_RBAR` Address 0xE000ED9C register bit assignments:
+![](img/armv7m_mpu/mpu_rbar.png)
+
+This register is used to control the starting address (aka the _base address_) that we want to cover with the region selected in `MPU_RNR`. For example, if we wanted to protect the address range starting at the beginning of a SRAM range (i.e `0x20000000`) using `REGION 0`, we would first set `MPU_RNR` to 0 and then set `MPU_RBAR` to `0x20000000`)
+
+To reduce the number of writes needed to change the configuration, the `VALID` bit can be set to `1` and the `REGION` field can be set to change the value of `MPU_RNR` and apply the write to a different region. For example, if we wanted to set `REGION 7` to have a base address of `0x20000000`, we could write `0x20000000 | 0x1 << 4 | 0x7 == 0x20000017` to `MPU_RBAR`. If we then read `MPU_RNR`, we would see its value is now `0x7`.
+
+**Caution** You may have noticed the address only occupies 27 bits and the bottom 5 bits were repurposed. What if we want to protect an address that starts with a base of `0x20000002`? You can't! The minimum region size is actually implementation defined but is usually 32 bytes. The actual size supported can be dyanmically determined by writing `0xFFFFFFE0` to `MPU_RBAR` and reading the value back. The number of trailing zeros will reveal the minimum supported alignment and consequently the minimum region size. For example, reading a value of `0xFFFFFF80` would reveal bits 5 and 6 can't be set so the minimum size supported is actually `2^7` (128) bytes.
```

Comment:
You can remove the word "actually" from most sentences, but it's especially wordy in this one.

> The minimum region size is ~~actually~~ implementation defined, ~~but~~ it is usually 32 bytes.

> The ~~actual~~ size supported can be ...

Reply:
Typo: dyamically -> dynamically

Reply:
Is there really no better way to find out the minimum region size!?!? I'd expect "the minimum region size is documented in your MCUs application notes" or something like that rather than this weird write / read.

Reply:
oh, yeah, totally can ... I think it's usually in the Cortex-M-specific reference manual. I'll add a note.

Reply:
Heh, so I see:

M0+ trm - "The MPU supports region sizes from 256-bytes to 4Gb, with 8-sub regions per region."
M3 trm - "The smallest permitted region size is 32B, corresponding to a SIZE value of 4."
M4 trm - nothing
M7 trm - nothing

So my guess would be M0+ is 256, M3,4,7 is always 32 but I'll just leave the note I added to just check the MCU manual :)